### PR TITLE
Allow proxy registration via agent role

### PR DESCRIPTION
**Pull request by ra2xfael**
To register agents [`proxy-registration`](https://docs.checkmk.com/latest/en/agent_linux.html#proxyregister) can be used on a separate machine, which is already fully provisioned (tls registered). This allows the role user to keep the firewall between ansible target and CheckMK server closed. The proxy-registration port only needs to be open between delegation target and server.

## Pull request type

- [x] Feature

## What is the current behavior?

Proxy registration is not possible using the agent ansible role. Regular registration needs to be used or proxy registration executed manually.

## What is the new behavior?

- `cmk-agent-ctl proxy-registration` is executed automatically on a delegate target and the encryption key is directly imported on the ansible target.
- The new variable `checkmk_agent_delegate_registration` is introduced by default. To set it `checkmk_agent_proxy_registration` can be used.

### CODE REVIEW

**Comment by ra2xfael** [comment]
I have read the CLA Document and I hereby sign the CLA or my organization already has a signed CLA.

**Comment by robin-checkmk** [comment]
@ra2xfael Please follow the instructions from the CLA bot, for us to be able to merge this PR (pending review).

**Comment by ra2xfael** [comment]
recheck

**Comment by ra2xfael** [comment]
I have read the CLA Document and I hereby sign the CLA or my organization already has a signed CLA.

**Comment by robin-checkmk** [comment]
Thanks for your contribution! We are pressed for time right now, but rest assured, this will be reviewed, as soon as we find the time. :v:

**Comment by robin-checkmk** [comment]
@ra2xfael in the meantime, you can look into the test failures and try to fix them. That will speed up the review process.

**Review by robin-checkmk — CHANGES_REQUESTED** [review]
You pushed changes, while I was reviewing, so I hope my review is not outdated. :)

Overall I really appreciate the quality of the pull request. There are just some minor changes necessary I think.

Also, it would be nice to keep an eye on idempotency, but it is not a hard requirement. I just know we have users, who have a keen eye on this. ;)

File: roles/agent/tasks/Linux.yml [hunk 1]
```diff
@@ -151,10 +151,78 @@
...
+    and checkmk_agent_tls | bool
+    and checkmk_agent_delegate_registration is defined
+    and (__checkmk_agent_auth is defined and __checkmk_agent_auth | length)
+    and not checkmk_agent_registration_server + '/' + checkmk_agent_registration_site in __checkmk_agent_registered_connections.stdout
+    and __checkmk_agent_proxy_registration_state is defined
+  changed_when: __checkmk_agent_import_result.rc == 0
+
+- name: Debug
+  debug:
+    var: checkmk_agent_delegate_registration
```
[the remark this answers is not in the document]
Comment (ra2xfael) [reply roles/agent/tasks/Linux.yml]: This is exactly the code I removed with the last change. So no problem there

File: roles/agent/defaults/main.yml [hunk 2]
```diff
@@ -30,6 +30,7 @@ checkmk_agent_force_install: 'false'
 checkmk_agent_prep_legacy: 'false'
 checkmk_agent_delegate_api_calls: 'localhost'
 checkmk_agent_delegate_download: "{{ inventory_hostname }}"
+checkmk_agent_delegate_registration: "{{ checkmk_agent_proxy_registration | default(omit) }}"
```
[the remark this answers is not in the document]
Comment (ra2xfael) [reply roles/agent/defaults/main.yml:33]: The change I try to add should only add one additional variable (to set the target of the delegation for the proxy-registration).
Because I was under the impression that all variables should be set via the defaults, I added `checkmk_agent_delegate_registration` which is then 'unset' by using omit. The variable `proxy_registration` is redundant.
I personally would remove the line in the defaults all together. This would deactivate proxy-registration by default.

**Comment by ra2xfael** [comment]
> You pushed changes, while I was reviewing, so I hope my review is not outdated. :)
> 
> Overall I really appreciate the quality of the pull request. There are just some minor changes necessary I think.
> 
> Also, it would be nice to keep an eye on idempotency, but it is not a hard requirement. I just know we have users, who have a keen eye on this. ;)

Thanks for the pull review @robin-checkmk !
I see the problems and would be keen to enhance my addition. There are just two points I need some clarification on.
1. Idempotency: The four tasks I added only run when `checkmk_agent_registration_server + '/' + checkmk_agent_registration_site in __checkmk_agent_registered_connections.stdout` is false. As I understand this, it is the case if the agent is not already connected with tls, so registration is still necessary. If tls registration is already done, then the tasks won't be executed. Idempotency should be guaranteed? 
2. Do all variables have to have a default? Can we just set a variable if delegation to do proxy-registration is necessary, otherwise if not set we assume it does not need to be done?

**Comment by robin-checkmk** [comment]
> * Idempotency: The four tasks I added only run when `checkmk_agent_registration_server + '/' + checkmk_agent_registration_site in __checkmk_agent_registered_connections.stdout` is false. As I understand this, it is the case if the agent is not already connected with tls, so registration is still necessary. If tls registration is already done, then the tasks won't be executed. Idempotency should be guaranteed?

In my quick test, it was not idempotent, but I would need to double-check. If it is idempotent on your end, it might be a local issue on my machine.

> * Do all variables have to have a default? Can we just set a variable if delegation to do proxy-registration is necessary, otherwise if not set we assume it does not need to be done?

Let's discuss this in the review comment above.

File: roles/agent/defaults/main.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (robin-checkmk) [reply roles/agent/defaults/main.yml:33]: > Two variables with a boolean switch and the delegation target seems pretty solid and if that's what we need to do I am fine with that, although I prefer one variable that can be unset.

I like the `solid` part if this. :)

> A third option could be checking whether the registration target is localhost (the default), which would cause regular registration, or a different value which than could trigger delegated proxy-registration.
This uses only one variable which is easier and more transparent and use the agent as intended by the app developers.

My problem with this is, that it is somewhat 'magic'. The logic would do a normal registration, if the task is delegated to `{{ inventory_hostname }}`, but a proxy registration, if it has another value. It seems like an elegant thing to do, but I believe it is better to be explicit, than implicit (which is actually one of the guiding principles of Ansible).


TL;DR: Let's go with the two variable approach, as it seems most robust. We can always simplify it, if we figure out a better way.

**Comment by robin-checkmk** [comment]
@ra2xfael I wanted to rebase your changes on `devel` and add some final touches to wording, but apparently GitHub did not like that too much. So now a lot of changes pop up, that should already be in `devel`. Maybe you can double-check the files you modified in the first place and if everything looks fine, I would go ahead and merge this.

**Review by robin-checkmk — APPROVED** [review]

**Comment by ra2xfael** [comment]
Hi @robin-checkmk, everything looks fine. The changes I made have been retained, there should be no problem with merging
